dap_aria_mapping/pipeline/semantic_taxonomy/make_taxonomies.py

```diff
@@ -0,0 +1,178 @@
+        " further clustering"
+    )
+    cluster_outputs_c, _ = run_clustering_generators(
+        CLUSTER_CONFIGS["centroids"], embeddings, embeddings_2d=embeddings_2d
+    )
+
+    logger.info("Running clustering generators - Done")
+
+    logger.info("Adjusting outputs from generator objects")
+    for output_dict in cluster_outputs_c:
+        for k, v in output_dict["labels"].items():
+            output_dict["labels"][k] = v[::-1]
+        output_dict["silhouette"] = output_dict["silhouette"][::-1]
+
+    cluster_outputs_f = []
+    for group in ["sklearn.cluster._kmeans", "sklearn.cluster._agglomerative"]:
+        dict_group = {
+            "labels": defaultdict(list),
+            "model": [],
+            "silhouette": [],
+            "centroid_params": None,
+        }
+
+        cluster_op = [
+            x for x in cluster_outputs_f_ if x["model"][0].__module__ == group
+        ]
+        for clust in cluster_op:
+            for k, v in clust["labels"].items():
+                dict_group["labels"][k].append(v[0])
+            dict_group["model"].append(
+                "_".join(
+                    [
+                        clust["model"][0].__module__.replace(".", ""),
+                        str(clust["model"][0].get_params()["n_clusters"]),
+                    ]
+                )
+            )
+            dict_group["silhouette"].append(clust["silhouette"][0])
+        cluster_outputs_f.append(dict_group)
+
+    logger.info("Running dictionary & table exports")
+    outputs_ls = [
+        cluster_outputs_s[3],
+        cluster_outputs_s[7],
+        cluster_outputs_simb[-1],
+        cluster_outputs_f[0],
+        cluster_outputs_f[1],
+        cluster_outputs_d,
+        cluster_outputs_c[-1],
+    ]
+
+    outputs_df = []
+    s3 = boto3.client("s3")
+
+    for dict_, df_config in zip(outputs_ls, df_configs):
+        name_export = df_config["name"]
+        s3.put_object(
+            Body=pickle.dumps(dict_),
+            Bucket=BUCKET_NAME,
+            Key=f"outputs/semantic_taxonomy/raw_outputs/{name_export}.pkl",
+        )
+
+        df = make_dataframe(dict_, df_config["label"], df_config["cumulative"])
+
+        if name_export == "agglom_dendrogram":
+            depth_dendrogram = df.shape[1]
+            levels = [int(depth_dendrogram * i) for i in [0.1, 0.3, 0.5, 0.7, 0.9]]
+            df = df.iloc[:, levels]
+
+        outputs_df.append(df)
+
+    logger.info("Creating meta_cluster object")
+    meta_cluster_df = pd.concat(outputs_df, axis=1)
+
+    logger.info("Running cooccurrence matrix export")
+    labels = [x["name"] for x in df_configs]
+    for label, df in zip(labels + ["meta_cluster"], outputs_df + [meta_cluster_df]):
+        logger.info(f"Running export routine for {label}")
+        df = (
+            df.rename(
+                columns={k: "Level_{}".format(v + 1) for v, k in enumerate(df.columns)}
+            )
+            .reset_index()
```

Comment:
do you need to reset the index, just to name it, and then set it back? I think you can just do df.index.name = 'Entity'

Reply:
Yes, but I don't know of any pipe-like way of doing this. I'll rather make it more cumbersome and keep it all within a single call to the df, rather than separate calls. Individual preference I suppose.